### PR TITLE
DOC: draft of a full explanation of norm + colormap interactions

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

This is the start of the documentation I said I would write on the call about float issues in resampling / normalizing / colormapping images, but I got distracted making sure that we had this well documented.

The text at https://matplotlib.org/3.3.1/tutorials/colors/colormapnorms.html is correct, but terse.  The code at https://matplotlib.org/3.3.1/gallery/userdemo/colormap_normalizations.html is almost the same but even less text.  https://matplotlib.org/3.3.1/tutorials/colors/colormaps.html is mostly about how to pick the right color map, and https://matplotlib.org/3.3.1/tutorials/colors/colormap-manipulation.html is about working with colormap objects once you have them.

I think the second one should probably go away and the first should link back to this new text.

### CODE REVIEW

File: doc/api/image_api.rst [hunk 1]
```diff
@@ -3,6 +3,172 @@
...
+   two Real Numbers there is an infinite number of Real numbers, hence
+   the floating point numbers and computation expressed in a computer
+   are an approximation of the Real Numbers.
+2. The absolute distance between adjacent floating point numbers
+   scales with the magnitude, while the relative distance remains
+   the same.  This is a consequence of the implementation of IEEE
+   floats.
+3. During computation results are rounded to the nearest
+   represent-able value.  Working with numbers that are either almost
+   identical or vastly different orders of magnitude exasperates the
```
Comment (jpivarski) [inline doc/api/image_api.rst:84]: "exasperates" → "exaggerates"? The former implies that the numbers lose their patience.

File: doc/api/image_api.rst [hunk 2]
```diff
@@ -3,6 +3,172 @@
...
+  data_min, data_max = data.min, data.max
+  # scale to [.1, .9]
+  rescaled = .1 + .8 * (data - data_min) / (data_max - data_min)
+  # get the correct number of pixels
+  resampled = resample(scaled)
+  # scale back to original data range
+  scaled = (resampled - .1) * (data_max - data_min)  + data_min
+
+For "most" user data is OK, but can fail in interesting ways.  First,
+if range of the input data is large, but the range the user actually
```
Comment (jpivarski) [inline doc/api/image_api.rst:111]: "if the range"

**Comment by jklymak** [comment]
So far, it looks like this includes pretty useful info at good places in the docs. 

However, I suggest being more colloquial.  I guess you could describe a colormap as a 1-D path through R^3 space, but maybe "a list of N colors specified as RGB" will be more accessible?  It took me a while to understand what you were trying to say, and my undergrad degree was Mathematics and Physics, albeit a few years ago ;-)

File: doc/api/colors_api.rst [hunk 3]
```diff
@@ -11,39 +11,157 @@ The Color :ref:`tutorials <tutorials-colors>` and :ref:`examples
...
+Normalization and Colormapping
+------------------------------
+
+Some `~.artists.Artist` classes can map an array of input data to RGBA
+values, (ex `~.axes.Axes.scatter` or `~.axes.Axes.imshow`).  The
+machinery for this is implemented via the `~.cm.ScalarMappable` base
+class in `~.cm` and the `~.Normalize` and `~.Colormap` classes in
+`~.colors` (this module).
+
+At the core, colormapping is going from a scalar value to a RGB tuple
```
Comment (jklymak) [inline doc/api/colors_api.rst:46]: I still feel this is more complicated than it needs to be, and not even particularly true if you want to be mathematically rigorous; colormaps are finite discrete ordered points in R^3, not a smooth curve or path, so I find taking about it like this distracting.   The first step is normalizing the data between vmin and vmax to the range 0-1, perhaps using a non-linear function, and secondly, using a linear lookup table to find the RGBA color closest to that value between 0 and 1.

Comment (jklymak) [reply doc/api/colors_api.rst:46]: I always like code examples to make the abstract concrete:
```
A color mapping from floating point data values to colors takes place in two steps.  First the data is normalized between 0 and 1.  This requires a `vmin` and `vmax` the define which values are mapped to 0 and 1, respectively, and it requires a function that defines the mapping between these two values.    The simplest case is a linear map ::
    
		import matplotlib.colors as mcolors
		norm = mcolors.Normalize(vmin=100, vmax=300) 
		norm([100, 200, 300])  # returns 0.0, 0.5, 1.0

but a logarithmic norm is also possible ::

    		norm = mcolors.LogNorm(vmin=10, vmax=1000)
		norm([10, 100, 1000])  # returns 0.0, 0.5, 1.0

Once the data has been normalized between 0 and 1, it can be passed to a colormap to return RGB(A) values ::
		
		cmap = cm.viridis
		cmap([0.0, 0.5, 1.0])  
	      #  returns RGBA array with the first three colors of viridis colormap.
	      # array([[0.267004, 0.004874, 0.329415, 1.      ],
             # [0.127568, 0.566949, 0.550556, 1.      ],
             # [0.993248, 0.906157, 0.143936, 1.      ]])
```

Comment (tacaswell) [reply doc/api/colors_api.rst:46]: We use discrete colormaps as an implementation detail (both for performance and because we are targeting 8bit RGB which is inherently discrete), but if we had infinite precision floats and an analytic function for the colormap the whole system would still hold together.

I think the correct solution is to include both versions :)

File: doc/api/image_api.rst [hunk 4]
```diff
@@ -3,6 +3,172 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Floating point numbers, despite being ubiquitous, are not fully
+understood by most practitioners.  For a through review of how
+floating point numbers work see `Goldberg, ACM Computing
+Surveys (1991) 10.1145/103162.103163
+<https://doi.org/10.1145/103162.103163>`__ or the `IEEE Standard for
+Floating Point Arithmetic ( IEEE std 754) 10.1109/IEEESTD.2008.4610935
+<https://doi.org/10.1109/IEEESTD.2008.4610935>`__ (both behind
+paywalls).  For the purposes of this discussion we need to know:
```
[the remark this answers is not in the document]
Comment (tacaswell) [reply doc/api/image_api.rst:70]: > Maybe you asked for help on some forum and got pointed to a long article with lots of formulas that didn’t seem to help with your problem.

Which is referring to the Goldberg article I cite

[the remark this answers is not in the document]
Comment (QuLogic) [reply doc/api/image_api.rst:70]: But not behind a paywall.

File: doc/api/colors_api.rst [hunk 5]
```diff
@@ -11,39 +11,157 @@ The Color :ref:`tutorials <tutorials-colors>` and :ref:`examples
...
+(formally :math:`f(x) : ℝ^1 \rightarrow ℝ^3`).  To effectively
+communicate through the color we want pick a :ref:`colormap suited to
+the data <sphx_glr_tutorials_colors_colormaps.py>` but in general
+"good" colormaps smoothly and continuously change their RGB values as
+a function of the input data.  By looking at the RGB values as we go
+through the full range of the user data (e.g. a colorbar) we can trace
```
[the remark this answers is not in the document]
Comment (story645) [reply doc/api/colors_api.rst:52]: Also the "good" in parenthesis is probably a stand in for some precise term and otherwise not true for categorical data and questionable on discrete data.

**Comment by tacaswell** [comment]
the colors_api.rst file is probably ready for a re-review, but I have not come back to the image_api.rst text yet.

File: doc/api/colors_api.rst [hunk 6]
```diff
@@ -14,42 +14,307 @@
...
+
+   get_named_colors_mapping
+
+
+.. _norms_and_colormaps:
+
+Normalization and Colormapping of Continuous Data
+-------------------------------------------------
+
+Some `~.artist.Artist` classes can map an array of input data to RGBA
```
Comment (jklymak) [inline doc/api/colors_api.rst:43]: There is a TOC aspect to this file, and then you've added an explanation.  a) not 100% sure the explanation belongs here, versus an intermediate/advanced tutorial.  Most long-form explanations are tutorials currently (i.e. transform stack).  I'm not adamantly advocating for that, but it would be more consistent with current practice.  It would also make your examples easier to write.  

Regardless, this large narrative in the middle of the TOC makes the TOC very hard to parse.  At the very least, this should go at the end?

File: doc/api/colors_api.rst [hunk 7]
```diff
@@ -14,42 +14,307 @@
...
+values, (ex `~.axes.Axes.scatter` or `~.axes.Axes.imshow`).  The
+machinery for this is implemented via the `~.cm.ScalarMappable` base
+class in `~.cm` and the `~.Normalize` and `~.Colormap` classes in
+`~.colors` (this module).
+
+At the core, colormapping is going from a scalar value to a RGB tuple (formally
+:math:`f(x) : ℝ^1 \rightarrow ℝ^3`).  To effectively communicate through the
```
Comment (jklymak) [inline doc/api/colors_api.rst:50]:
```suggestion
At the core, colormapping is going from a scalar value to a RGB tuple (formally
:math:`f(x) : ℝ^1 \rightarrow ℝ_[0,1]^1  \rightarrow ℝ^3`).  To effectively communicate through the
```

File: doc/api/colors_api.rst [hunk 8]
```diff
@@ -14,42 +14,307 @@
...
+color we want pick a :ref:`colormap suited to the data
+<sphx_glr_tutorials_colors_colormaps.py>`.  For continuous data types [#f1]_ a
```
Comment (jklymak) [inline doc/api/colors_api.rst:52]: The outline here is a little non-linear, and seems to take a while before it settles into an order.  
- mapping
- choosing colormap
- mapping
- norm
- norm
- colormapping 

I think this could all be three paragraphs, or maybe 5 if you want to be expansive:

1) overview
2) norm step
3) colormap step

If you plan to talk about the "Curve through  space"  I wonder if a diagram would make this much clearer.  Expressing this with math formalisms without diagrams is going to lose your audience.
